Feat/add hooks repository

charley-geoffroy opened the pull request:
**Update :**

J'ai apporté les modifications que tu souhaitais, effectivement il est plus logique de supprimer le repo si la commande ne passe pas . 

J'ai aussi implémenté un **client debug grpc**, qui me semble essentielle pour que tout le monde puisse tester.

Commit:
feat: add hook to repository

Commit:
fix: give right to execute hook

Commit:
fix: lien symbolique hooks

Commit:
fix: cleanup repo + client debug server

Commit:
feat/ client grpc ask repo name

mmoreiradj commented on git-repo-manager/src/service.rs:
```diff
@@ -49,6 +51,29 @@ impl GitRepoManager for GitRepoManagerService {
+        if Command::new("sh")
+            .arg("-c")
+            .arg(format!(
+                "ln -s /home/git/bin/post-receive {}/.git/hooks/post-receive",
```
ce serait bien de faire du /home/git/bin/ une variable d'environnement comme ca si on veux tester en local on peux modifier la source, par exemple $HOME/git-controller/hooks/

mmoreiradj replied:
En plus tu le copie pas au bon endroit. C'est {}/hooks le chemin

mmoreiradj commented on git-repo-manager/src/service.rs:
```diff
@@ -49,6 +51,29 @@ impl GitRepoManager for GitRepoManagerService {
             return Err(Status::unknown("Failed initializing repository"));
         }
 
+        // add post-receive hook : le lien symbolique peut être remplacé
```
franglais

mmoreiradj commented on git-repo-manager/src/service.rs:
```diff
@@ -49,6 +51,29 @@ impl GitRepoManager for GitRepoManagerService {
+            .output()
+            .is_err()
+        {
+         // clean up the repository if it fails rollback and return error otherwise continue
```
run un petit cargo fmt

mmoreiradj commented on grpc-client-debug/README.md:
````diff
@@ -0,0 +1,12 @@
+To run the example, first make sur to have started the gRPC server then run the following command:
+
+```bash
+cargo build
````
pas besoin de build, cargo run le fait déjà

mmoreiradj commented on git-repo-manager/src/hooks/post-receive:
pourquoi pas mettre ce fichier à la racine du projet genre git-controller/build-packs/hooks/...

Commit:
update: path changed to variable

Commit:
fix: githook path as .env virables

mmoreiradj commented on git-repo-manager/README.md:
retours a la ligne en trop à la fin et a la ligne 13

mmoreiradj commented on git-repo-manager/src/service.rs:
```diff
@@ -23,8 +24,13 @@ impl GitRepoManager for GitRepoManagerService {
         &self,
         request: Request<RepositoryRequest>,
     ) -> GitSproutResult<RepositoryResponse> {
+
```
run un cargo fmt (on fera un CI au plus vite)

mmoreiradj commented on git-repo-manager/src/service.rs:
```diff
@@ -23,8 +24,13 @@ impl GitRepoManager for GitRepoManagerService {
         let request_data = request.into_inner();
 
+        // Create the full path to the repository
+
```
whitespace inutile ( non détecté par cargo fmt attention)

mmoreiradj commented on git-repo-manager/src/service.rs:
```diff
@@ -49,6 +55,31 @@ impl GitRepoManager for GitRepoManagerService {
         }
 
+        // add post-receive hook : le lien symbolique peut être remplacé
+        if Command::new("sh")
```
Je n'ai pas l'impression que ce check fonctionne, si j'enlève le post-receive de hooks pour le tester ou alors que je mets un chemin qui n'existe pas, le repo est quand même créé.

Commit:
Fix: verify if hook file exists

Signed-off-by: Charley <charley.geoffroy@etu.umontpellier.fr>

Martin-Moreira-de-jesus commented on git-repo-manager/.env:
```diff
@@ -1,4 +1,4 @@
 GITSPROUT_GIT_REPOSITORY_BASE_PATH=./srv
 GITSPROUT_SERVER_PORT=50051
-GITSPROUT_GITHOOK_BASE_PATH=~/polytech/woa/git-controller/
+GITSPROUT_GITHOOK_BASE_PATH= /home/charley/polytech/woa/git-controller
```
évites de push ca :/

mmoreiradj replied:
Dont push this. You should instead push **/path/to/githooks** par exemple

Martin-Moreira-de-jesus commented on git-repo-manager/src/service.rs:
```diff
@@ -77,7 +87,7 @@ impl GitRepoManager for GitRepoManagerService {
             }
 
-            return Err(Status::unknown("Failed adding post-receive hook"));
+            return Err(Status::unknown("Failed hooks path to post-receive file doesn't exist"));
```
Not english, "Failed linking post-receive hook, file does not exist" is a more descriptive message

mmoreiradj replied:
Not english, "Failed linking post-receive hook, file does not exist" is a more descriptive message

mmoreiradj commented on git-repo-manager/src/service.rs:
```diff
@@ -55,18 +54,29 @@ impl GitRepoManager for GitRepoManagerService {
-                self.config.githook_base_path.clone(),
-                new_repo_path.clone(),
+        let hook_path = format!(
+            "{}/hooks/post-receive",
```
I think it should rather be "{}/post-receive", that way we can put the post-receive anywhere without having to add hooks, what do you think ?

mmoreiradj commented on git-repo-manager/src/service.rs:
cargo fmt

mmoreiradj commented on git-repo-manager/src/service.rs:
```diff
@@ -55,18 +54,29 @@ impl GitRepoManager for GitRepoManagerService {
-        {
+        let hook_path_exists = Path::new(&hook_path).exists();
+
+        if hook_path_exists {
```
Put line 62 directly in the if at line 64 ?

```if Path::new(&hook_path).exists(); {}``` ?

mmoreiradj commented on git-repo-manager/src/service.rs:
```diff
@@ -55,18 +54,29 @@ impl GitRepoManager for GitRepoManagerService {
+        let hook_path_exists = Path::new(&hook_path).exists();
+
+        if hook_path_exists {
+            // Le chemin existe, vous pouvez lancer la commande
```
Comments in english please

mmoreiradj reviewed (changes requested):
Minor changes, mostly form, but everything else works ! :clap:

Commit:
fix: cargo fmt update

Commit:
fix: cargo clippy check

Commit:
docs(repo-manager): fix comments

Commit:
refactor(repo-manager): move hook logic

Move check if hook exist at the beginning of the function.
Remove clean up in case of fail since `ln -s` cannot fail.

Commit:
fix(repo-manager): normalize env vars

Create constant files to store environment variables keys and normalize names.
Panic with error message when env var is missing.
Remove personal information from env var and put absolute paths.
Rename git hook environment variable key.

Commit:
fix(repo-manager): remote ununsed file
